Update CSR docs with expirationSeconds field details (#28070)
kubernetes/website: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md

the `spec.request` field. The CertificateSigningRequest denotes the signer (the
recipient that the request is being made to) using the `spec.signerName` field.
Note that `spec.signerName` is a required key after API version `certificates.k8s.io/v1`.

> The client may optionally set the `spec.durationHint` to request a particular

Reviewer: It seems like this needs to change to reflect the field renaming to `.spec.expirationSeconds`

> lifetime for the issued certificate.

Reviewer: For this first mention of the field:
- I'd mention it's only available in Kubernetes v1.22 and later
- I'd also tell people what the type of the field is. They might for example think it takes a number of seconds or a number of days, or more likely be frustrated because both could be valid and they're not sure.

Reviewer 2: yes to noting the version this is supported in

API convention is to include the unit in the field name and make it an integer, not a string (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units)

Reviewer: I guess I'd like this to be called `durationSecondsHint` or similar, then (and be an integer).

Reviewer: If the field name is already decided, then instead we must document what units it uses. Ideally the API reference for CertificateSigningRequest will use the same, clear working.

Reviewer 2: final field name is expirationSeconds

* Approved requests: automatically deleted after 1 hour
* Denied requests: automatically deleted after 1 hour
* Failed requests: automatically deleted after 1 hour
* Pending requests: automatically deleted after 24 hours

> * All requests: automatically deleted after the issued certificate has expired

Author: The existing docs are incorrect and incomplete.

> * All requests: automatically deleted after the issued certificate has expired

Reviewer 2: I'd suggest splitting the corrections to current docs to a PR to `master` and merging it now

additional certificates; for example, this might be the certificate plus
intermediates to be presented during TLS handshakes.

The PKCS#10 signing request format doesn't allow to specify a certificate
The PKCS#10 signing request format does not allow to specify a certificate
expiration or lifetime. The expiration or lifetime therefore has to be set

> through e.g. an annotation on the CSR object. While it's theoretically

Reviewer: nit: PKCS#10 as a format does allow specifying this (via an extension); the problem is that there's not standard OID for recording the expiration date or lifetime. (I think we're saying that this separate mechanism makes more sense for Kubernetes than requiring users to learn how to use some custom CNCF OID for that, with all the ASN.1 fun that would entail - and I agree!

(the `--cluster-signing-duration` command-line flag of the kube-controller-manager)
as the default when no `spec.durationHint` is specified.  When `spec.durationHint`
is specified, the minimum of `spec.durationHint` and `ClusterSigningDuration` is
used.

{{< note >}}

> The `spec.durationHint` field was added in Kubernetes v1.22.  Consequently, earlier versions of Kubernetes will not honor this field.

Reviewer: ```suggestion
The `spec.durationHint` field was added in Kubernetes v1.22. Earlier versions of Kubernetes do not honor this field.
```
What happens if you try to set that field in a CSR submitted to an older API server?

Author: Like all new fields added in newer Kubernetes versions, older API servers will silently drop the field.

Reviewer: > Like all new fields added in newer Kubernetes versions, older API servers will silently drop the field.

I recommend explaining that in the text: we shouldn't assume readers are closely familiar with Kubernetes API versioning rules.

    1. Permitted subjects - any
    1. Permitted x509 extensions - honors subjectAltName and key usage extensions and discards other extensions.
    1. Permitted key usages - any
    1. Expiration/certificate lifetime - set to the minimum of either the `--cluster-signing-duration` option for the
       kube-controller-manager implementation of this signer or the `spec.durationHint` field of the CSR object if specified.
    1. CA bit allowed/disallowed - not allowed.

{{< note >}}
Failures for all of these are only reported in kube-controller-manager logs.
{{< /note >}}

{{< note >}}

> The `spec.durationHint` field was added in Kubernetes v1.22.  Consequently, earlier versions of Kubernetes will not honor this field.

Reviewer: ```suggestion
The `spec.durationHint` field was added in Kubernetes v1.22.  Earlier versions of Kubernetes do not honor this field.
```

were marked as approved.
{{< /note >}}

{{< note >}}

> The `spec.durationHint` field was added in Kubernetes v1.22.  Consequently, earlier versions of Kubernetes will not honor this field.

Reviewer: ```suggestion
The `spec.durationHint` field was added in Kubernetes v1.22.  Earlier versions of Kubernetes do not honor this field.
```
